File: codeWriting1.java
import java.util.ArrayList;
import java.util.Scanner;

public class Solution {
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);

		int num_caches = Integer.parseInt(scanner.nextLine());
		String[] caches_capacities = scanner.nextLine().split(" ");

		Cache[] caches_list = new Cache[num_caches];
		for (int i = 0; i < caches_capacities.length; i++) {
			int size = Integer.parseInt(caches_capacities[i]);
			caches_list[i] = new Cache(size);
		}
		while (true) {
			String input = scanner.nextLine();
			if (input.contains("RANGE")) {
				processRange(input, caches_list);
			} else if (input.contains("ADDR")) {
				processAddr(input, caches_list);
			} else if (input.contains("STAT")) {
				processStat(caches_list);
			} else if (input.contains("END")) {
				break;
			}
		}
		scanner.close();
	}

	private static void processRange(String input, Cache[] caches_list) {
		String[] data = input.split(" ");
		int b = Integer.parseInt(data[1]);
		int y = Integer.parseInt(data[2]);
		int n = Integer.parseInt(data[3]);
		for (int k = 0; k < n; k++) {
			int data_reference = b + (y * k);
			addDataReference(data_reference, caches_list);
		}
	}

	private static void processAddr(String input, Cache[] caches_list) {
		String[] data = input.split(" ");
		int data_reference = Integer.parseInt(data[1]);
		addDataReference(data_reference, caches_list);
	}

	private static void processStat(Cache[] caches_list) {
		for (int i = 0; i < caches_list.length; i++) {
			if (i < caches_list.length - 1){
				System.out.print(caches_list[i].getMiss());
				System.out.print(" ");
			} else {
				System.out.println(caches_list[i].getMiss());
			}
			caches_list[i].resetMiss();
		}
	}

	private static void addDataReference(int data_reference, Cache[] caches_list) {
		for (Cache cache : caches_list) {
			cache.check(data_reference);
		}
	}
}

class Cache {
	int miss;
	int cache_size;
	ArrayList<Integer> cache_list;

	public Cache(int cache_size) {
		this.miss = 0;
		this.cache_size = cache_size;
		this.cache_list = new ArrayList<>();
	}

	public void check(Integer data) {
		if (cache_list.contains(data)) {
			cache_list.remove(data);
			cache_list.add(data);
		} else {
			miss++;
			cache_list.add(data);
			if (cache_list.size() > cache_size) {
				cache_list.remove(0);
			}
		}
	}

	public int getMiss() {
		return miss;
	}

	public void resetMiss() {
		this.miss = 0;
	}
}

File: codeWriting1.py
class Cache:
    def __init__(self, capacity):
        self.size = capacity
        self.cache_list = []
        self.miss = 0

    def check(self, data):
        if data in self.cache_list:
            self.cache_list.remove(data)
            self.cache_list.append(data)
        else:
            self.miss += 1
            self.cache_list.append(data)
            if len(self.cache_list) > self.size:
                self.cache_list.pop(0)
        pass

    def getMiss(self):
        return self.miss

    def resetMiss(self):
        self.miss = 0
        pass


def addDataReference(data, caches):
    for cache in caches:
        cache.check(data)
    pass


def processRange(line, caches):
    line = line.split(" ")
    b = int(line[1])
    y = int(line[2])
    n = int(line[3])
    for k in range(0, n):
        data = b + (y * k)
        addDataReference(data, caches)
    pass


def processAddr(line, caches):
    line = line.split(" ")
    data = int(line[1])
    addDataReference(data, caches)
    pass


def processStat(caches):
    misses = []
    for cache in caches:
        misses.append(cache.getMiss())
        cache.resetMiss()

    misses = str(misses).replace("[", "").replace("]", "").replace(", ", " ")
    print(misses)
    pass


if __name__ == '__main__':
    num_caches = int(input())
    cache_capacity_list = input().split(" ")

    caches_list = []
    for size in cache_capacity_list:
        caches_list.append(Cache(int(size)))

    for input_line in iter(input, ""):
        if "RANGE" in input_line:
            processRange(input_line, caches_list)
        elif "ADDR" in input_line:
            processAddr(input_line, caches_list)
        elif "STAT" in input_line:
            processStat(caches_list)
        elif "END" in input_line:
            break
